File: backend/auth/signin.ts
import { Router } from "express";
import type { Request, Response } from "express";
import prisma from "../../lib/prisma.js";
import { verifyPassword } from "../../lib/auth.js";
import { createSession } from "./helpers.js";

const router = Router();

router.post("/", async (req: Request, res: Response) => {
    try {
        const { email, password } = req.body as {
            email?: string;
            password?: string;
        };

        if (!email || !password) {
            res.status(400).json({ error: "Email and password are required" });
            return;
        }

        const user = await prisma.user.findUnique({ where: { email } });
        if (!user || !user.password_hash) {
            res.status(401).json({ error: "Invalid email or password" });
            return;
        }

        if (!user.is_verified) {
            res.status(403).json({ status: "pending_verification", email });
            return;
        }

        const valid = await verifyPassword(password, user.password_hash);
        if (!valid) {
            res.status(401).json({ error: "Invalid email or password" });
            return;
        }

        const accessToken = await createSession(user.id, res, req);
        res.json({
            accessToken,
            user: { id: user.id, email: user.email, name: user.name, avatar_url: user.avatar_url },
        });
    } catch (err) {
        console.error("[signin]", err);
        res.status(500).json({ error: "Internal server error" });
    }
});

export default router;

File: backend/auth-service/app/routers/signin.py
from fastapi import APIRouter, Request, Response
from pydantic import BaseModel, EmailStr

from app.database import get_pool
from app.lib.auth import verify_password
from app.routers._helpers import create_session

router = APIRouter()


class SigninRequest(BaseModel):
    email: EmailStr
    password: str


@router.post("/signin")
async def signin(body: SigninRequest, request: Request, response: Response):
    try:
        pool = get_pool()

        user = await pool.fetchrow(
            "SELECT id, email, name, avatar_url, password_hash, is_verified FROM users WHERE email = $1",
            body.email,
        )

        if not user or not user["password_hash"]:
            response.status_code = 401
            return {"error": "Invalid email or password"}

        if not user["is_verified"]:
            response.status_code = 403
            return {"status": "pending_verification", "email": body.email}

        if not verify_password(body.password, user["password_hash"]):
            response.status_code = 401
            return {"error": "Invalid email or password"}

        access_token = await create_session(str(user["id"]), response, request)

        return {
            "accessToken": access_token,
            "user": {
                "id": str(user["id"]),
                "email": user["email"],
                "name": user["name"],
                "avatar_url": user["avatar_url"],
            },
        }

    except Exception as e:
        print(f"[signin] {e}")
        response.status_code = 500
        return {"error": "Internal server error"}
